File: client_global.go
package groundcover

import (
	"context"
	"sync/atomic"
	"time"
)

// globalRecloseTimeout bounds the background teardown of a previous global
// client when Init is called more than once.
const globalRecloseTimeout = 5 * time.Second

// global holds the package-level default client used by the top-level
// functions. It is the single intentional package-level mutable global in
// the SDK; all other state is per-Client. It starts as a no-op client so the
// package functions are safe to call before Init.
//
//nolint:gochecknoglobals // intentional package-level default client
var global atomic.Pointer[Client]

// disabledClient is a shared no-op client used until Init succeeds.
//
//nolint:gochecknoglobals // immutable sentinel for the uninitialized global
var disabledClient = func() *Client {
	c, _ := newClient(Config{Disabled: true}, nil)
	return c
}()

// currentGlobal returns the active global client, or the no-op client if Init
// has not been called.
func currentGlobal() *Client {
	if c := global.Load(); c != nil {
		return c
	}
	return disabledClient
}

// Init configures the package-level default client. Calling it again replaces
// the previous default and tears the old one down in the background (a bounded,
// best-effort Close) so its worker goroutine does not leak. Init never blocks on
// that teardown.
func Init(cfg Config) error {
	c, err := New(cfg)
	if err != nil {
		return err
	}
	prev := global.Swap(c)
	if prev != nil && prev != c {
		go func() {
			ctx, cancel := context.WithTimeout(context.Background(), globalRecloseTimeout)
			defer cancel()
			_ = prev.Close(ctx)
		}()
	}
	return nil
}

// CaptureError captures err using the package-level client.
func CaptureError(ctx context.Context, err error, opts ...Option) {
	currentGlobal().CaptureError(ctx, err, opts...)
}

// CaptureMessage captures a non-error notice using the package-level client.
func CaptureMessage(ctx context.Context, msg string, level Level, opts ...Option) {
	currentGlobal().CaptureMessage(ctx, msg, level, opts...)
}

// Recover captures a panic (then re-raises it) using the package-level client.
// Use it deferred: defer groundcover.Recover(ctx).
func Recover(ctx context.Context) {
	r := recover()
	if r == nil {
		return
	}
	c := currentGlobal()
	if !c.disabled {
		c.CaptureRecovered(ctx, r)
		// Detached best-effort flush before re-raising the panic.
		flushCtx, cancel := context.WithTimeout(context.Background(), panicFlushTimeout)
		_ = c.Flush(flushCtx) //nolint:contextcheck // deliberate detached best-effort flush before re-raise
		cancel()
	}
	panic(r)
}

// CaptureRecovered captures an already-recovered panic value without re-raising,
// using the package-level client.
func CaptureRecovered(ctx context.Context, recovered any, opts ...Option) {
	currentGlobal().CaptureRecovered(ctx, recovered, opts...)
}

// SetUser returns a context with the identity set, using the package-level client.
func SetUser(ctx context.Context, u User) context.Context {
	return currentGlobal().SetUser(ctx, u)
}

// WithScope applies fn to the request scope (mutating an existing scope in
// place), using the package-level client.
func WithScope(ctx context.Context, fn func(*Scope)) context.Context {
	return currentGlobal().WithScope(ctx, fn)
}

// WithIsolatedScope returns a context with a fresh, isolated copy of the current
// scope, using the package-level client. Middleware uses it at request boundaries.
func WithIsolatedScope(ctx context.Context) context.Context {
	return currentGlobal().WithIsolatedScope(ctx)
}

// Flush flushes the package-level client.
func Flush(ctx context.Context) error { return currentGlobal().Flush(ctx) }

// Close closes the package-level client.
func Close(ctx context.Context) error { return currentGlobal().Close(ctx) }

// FlushTimeout flushes the package-level client with a context bounded by d.
func FlushTimeout(d time.Duration) error { return currentGlobal().FlushTimeout(d) }

// CloseTimeout closes the package-level client with a context bounded by d.
func CloseTimeout(d time.Duration) error { return currentGlobal().CloseTimeout(d) }

// GlobalStats returns the package-level client's self-metrics snapshot. (The
// per-client accessor is the Client.Stats method; this avoids colliding with
// the Stats type at package scope.)
func GlobalStats() Stats { return currentGlobal().Stats() }

File: groundcover-python/src/groundcover/_client_global.py
"""The module-level default client (Sentry style).

This holds the single intentional module-level mutable global in the SDK; all
other state is per-Client. It starts as a no-op client so the module functions
are safe to call before init().
"""

from __future__ import annotations

import threading
from typing import Any, Callable, ContextManager, Optional, Union

from ._client import PANIC_FLUSH_TIMEOUT, Client
from ._config import Config
from ._event import Attributes
from ._internal import safeguard
from ._level import Level
from ._scope import Scope, isolated_scope
from ._stats import Stats
from ._user import User

GLOBAL_RECLOSE_TIMEOUT = 5.0
"""Bounds the background teardown of a previous global client when init() is
called more than once."""

_lock = threading.Lock()
_client: Optional[Client] = None
_disabled_client: Optional[Client] = None


def _noop_client() -> Client:
    """Return the shared no-op client used until init() succeeds."""
    global _disabled_client
    with _lock:
        if _disabled_client is None:
            _disabled_client = Client(Config(disabled=True))
        return _disabled_client


def current_global() -> Client:
    """Return the active global client, or the no-op client if init() has not
    been called."""
    c = _client
    if c is not None:
        return c
    return _noop_client()


def init(config: Optional[Config] = None, **kwargs: Any) -> None:
    """Configure the module-level default client.

    Accepts either a Config or Config field names as keyword arguments:

        groundcover.init(dsn="https://...", ingestion_key="...")

    Calling it again replaces the previous default and tears the old one down
    in the background (a bounded, best-effort close) so its worker thread does
    not linger. init() never blocks on that teardown. Raises MissingDSNError
    when the configuration is unusable.
    """
    global _client
    if config is None:
        config = Config(**kwargs)
    elif kwargs:
        raise TypeError("pass either a Config or keyword arguments, not both")

    c = Client(config)
    with _lock:
        prev = _client
        _client = c
    if prev is not None and prev is not c:
        threading.Thread(
            target=lambda: prev.close(GLOBAL_RECLOSE_TIMEOUT),
            daemon=True,
            name="groundcover-reclose",
        ).start()


def capture_error(
    error: Optional[BaseException],
    *,
    attributes: Optional[Attributes] = None,
    user: Optional[User] = None,
    level: Union[Level, str, None] = None,
    fingerprint: Optional[str] = None,
    title: Optional[str] = None,
) -> None:
    """Capture an exception using the module-level client."""
    current_global().capture_error(
        error,
        attributes=attributes,
        user=user,
        level=level,
        fingerprint=fingerprint,
        title=title,
    )


def capture_message(
    message: str,
    level: Union[Level, str] = Level.INFO,
    *,
    attributes: Optional[Attributes] = None,
    user: Optional[User] = None,
    fingerprint: Optional[str] = None,
    title: Optional[str] = None,
) -> None:
    """Capture a non-error notice using the module-level client."""
    current_global().capture_message(
        message,
        level,
        attributes=attributes,
        user=user,
        fingerprint=fingerprint,
        title=title,
    )


def capture_recovered(
    recovered: Any,
    *,
    attributes: Optional[Attributes] = None,
    user: Optional[User] = None,
    level: Union[Level, str, None] = None,
    fingerprint: Optional[str] = None,
    title: Optional[str] = None,
) -> None:
    """Capture an already-caught exception as unhandled, without re-raising,
    using the module-level client."""
    current_global().capture_recovered(
        recovered,
        attributes=attributes,
        user=user,
        level=level,
        fingerprint=fingerprint,
        title=title,
    )


class _GlobalRecoverContext:
    """Context manager backing the module-level recover(). It resolves the
    global client at exit time so a client installed after entering is still
    used."""

    def __enter__(self) -> None:
        return None

    def __exit__(self, exc_type: Any, exc: Any, tb: Any) -> bool:
        if exc is None or not isinstance(exc, Exception):
            return False
        c = current_global()
        if not c.disabled:

            def _do() -> None:
                c.capture_recovered(exc)
                # Detached best-effort flush before the exception re-raises.
                c.flush(PANIC_FLUSH_TIMEOUT)

            safeguard.do(_do, c._on_panic)
        return False


def recover() -> ContextManager[None]:
    """Return a context manager that captures an escaping exception (then
    re-raises it) using the module-level client:

        with groundcover.recover():
            do_risky_work()
    """
    return _GlobalRecoverContext()


def set_user(user: User) -> Scope:
    """Set the identity on the current request scope, using the module-level
    client."""
    return current_global().set_user(user)


def with_scope(fn: Optional[Callable[[Scope], None]]) -> Scope:
    """Apply fn to the current request scope (mutating an existing scope in
    place), using the module-level client."""
    return current_global().with_scope(fn)


def flush(timeout: Optional[float] = None) -> bool:
    """Flush the module-level client."""
    return current_global().flush(timeout)


def close(timeout: Optional[float] = None) -> bool:
    """Close the module-level client."""
    return current_global().close(timeout)


def global_stats() -> Stats:
    """Return the module-level client's self-metrics snapshot. (The
    per-client accessor is the Client.stats method; this avoids colliding
    with the Stats type at module scope.)"""
    return current_global().stats()


__all__ = [
    "capture_error",
    "capture_message",
    "capture_recovered",
    "close",
    "current_global",
    "flush",
    "global_stats",
    "init",
    "isolated_scope",
    "recover",
    "set_user",
    "with_scope",
]
